include/deal.II/grid/grid_tools.h

```diff
@@ -1727,6 +1727,9 @@ namespace GridTools
                      + cell->face(face)->vertex(2) + cell->face(face)->vertex(3)) / 4.0;
               }
       }
+
+    // Make sure FEValues notices that the mesh has changed
+    triangulation.signals.post_refinement();
```

Comment:
I agree.

I'm going to note for the record that we do allow explicitly moving vertices by doing
```
  cell->vertex(3) = Point<dim>(1,2,3);
```
That is a historical wart that I'm not proud of, precisely because it leads to issue such as the one you're seeing -- the rest of the library will never know this happened.